state/workers.go

```diff
@@ -159,9 +170,29 @@ func (ws *workers) allManager(params WatchParams) *storeManager {
 		return newDeadStoreManager(errors.Trace(err))
 	}
 	// Note that StartWorker is idempotent if there's a race.
-	ws.StartWorker(allManagerWorker, func() (worker.Worker, error) {
-		return newStoreManager(newAllWatcherStateBacking(ws.state, params)), nil
-	})
+	if err := ws.StartWorker(allManagerWorker, func() (worker.Worker, error) {
+		// The allWatcher uses system state to load all the entities to watch.
+		// The system state is not ref counted like other states that are fetched
+		// from the state pool, so create a copy here and use that to guard against
+		// the possibility that the system state may be closed elsewhere.
+		// The state copy is closed then the store manager is released.
+		stCopy, err := Open(OpenParams{
```

Comment:
I think this is the wrong approach. Firstly, using Open causes a new txn watcher to be created for that state object.

We should be more careful about the lifetime of the workers that have dependent objects.